### ISSUE TITLE
Add and enforce style

### ISSUE THREAD

**Issue by JensHeinrich** [ansible-collections/community.crypto#359]
##### SUMMARY

Define coding style conventions and enforce them

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME

all

##### ADDITIONAL INFORMATION

Suggestions:
* Use [`ansible-lint`](https://github.com/ansible-community/ansible-lint) to style test playbooks (can be easily enforced using [`pre-commit`](https://pre-commit.com/)
* Use [`.pre-commit-config.yaml`](https://github.com/ansible-collections/community.general/blob/main/.pre-commit-config.yaml) and  [`CONTRIBUTING.md`](https://github.com/ansible-collections/community.general/blob/main/CONTRIBUTING.md) from [community.general](https://github.com/ansible-collections/community.general)

```yaml

```

**Comment by felixfontein** [comment]
Please note that the pre-commit-config.yaml from c.g does not do anything that `ansible-test sanity` (which we are using in CI) isn't already doing.

**Comment by JensHeinrich** [comment]
This is a valid point

Maybe an `ansible-test` pre-commit hook would be rather useful, but this would be upstream (as in `ansible/ansible`) I think.

Still setting up pre-commit is rather useful and enables contributers to rely on automation locally before having to create fixes upstream

**Comment by JensHeinrich** [comment]
Also I am still very much a fan of the `ansible-lint` style for playbooks as it makes them much easier to read

**Comment by MarkusTeufelberger** [comment]
I recently published some of my company's roles (https://github.com/mgit-at/ansible-collection-roles) and running `ansible-lint` in `pre-commit` is a significant burden on development speed due to the design of both tools (and some automagic performed by `ansible-lint`...).

I'm more than happy to have fast pre-commit hooks enabled, but looking at how much trouble `ansible-lint` is already giving me for a simple collection with only a handful of roles I'd rather have it run as GitHub Action and/or `manual` hook in `pre-commit`, definitely NOT as something that gets installed and used by default. There has to be some major overhaul and work done to make this tool combination usable.

I'd add https://github.com/adrienverge/yamllint and https://github.com/shellcheck-py/shellcheck-py to the hooks by the way - those are really fast and can be useful. Personally I'm partial to pre-commit hooks that actually rewrite files during the commit phase and I much prefer _not_ having stuff I wrote suddenly and automatically modified or re-written, but if we use https://pre-commit.ci/ their bot can take care of this (and also auto-updating the hooks).

**Comment by MarkusTeufelberger** [comment]
I think there are several layers of possible complexity in pre-commit hooks/checks that could be added here:

* line endings, whitespace etc. as in c.g. -- probably very fast and safe as a starting point (add to the repository, enable on https://pre-commit.ci/ etc.)
* yamllint/shellcheck -- somewhat opinionated, still very fast, might need some configuration files
* `ansible test` -- already runs in CI anyways(?), no idea about speed or actual usage + context required (likely needs to run on the whole repo, not only on changes)
* `ansible-lint` -- opinionated, takes (imho far too) long to start and currently will always scan everything everywhere which takes a while
* ...`molecule`? I mean, it could work, but that's the slowness of Ansible + adding a whole layer on top! Also likely additional requirements on the host machine besides just having Python and `pre-commit` installed - such as `vagrant`, `docker` etc...

We can easily start with the first or first + second points, but for `ansible test`, `ansible-lint` and `molecule` I'd like to see clear performance goals (e.g. "on a simple change all pre-commit hooks _have_ to finish in under a second wall time total"). I have my doubts that these are reachable. There might be some middle ground in using the `pre-commit` tool as a sort of "package/dependency manager" so it is easy for someone to develop and commit and then in the end also run the CI checks locally by running `pre-commit run --all-files --hook-stage manual` before pushing to GitHub to get some faster feedback.

**Comment by Ajpantuso** [comment]
Just my 2 cents after the discussion in `c.g` a few months ago:

- `ansible-test sanity` has final say on code style/convention so as mentioned any hooks should not conflict (adding an entrypoint to the linting portion of `sanity` in `ansible/ansible` is possible albeit gross).
- `ansible-lint` is not really relevant here, or in `c.g`, except maybe with a very limited ruleset for integration tests (though those end up being hacky for various reasons).

Overall I would focus on "is it parseable/interpretable" and basic code quality concerns for `pre-commit`. As for more nuanced style guidelines I would add links to Ansible's docs and add anything you feel is missing there. It might make sense however to have more specific guidelines related to security here that the other collections could draw from.

**Comment by felixfontein** [comment]
I don't mind having an **optional** pre-commit script (same as in community.general) that is not used in CI. The idea there is to catch some common problems that are easy to detect and fix, that are also caught by the regular CI.

Running ansible-test or ansible-lint in such a pre-commit script is probably not a good idea (see for example https://github.com/ansible-collections/community.general/pull/3562#issuecomment-945782637), not just because these tend to be slow, and in case of ansible-test, require a special directory setup for your git checkout.

On using ansible-lint: I haven't seen it really improve readability (except screaming at you when you have a task without `name:`). So I'm not sure how much use it will really be (except being a PITA by being really opinionated, and not always in a good way).

**Comment by JensHeinrich** [comment]
@MarkusTeufelberger TBH I have ansible-lint even as an asychronous tool in my editor, but I see the amount of problems it can make (initially) too; but after it has been setup and the repository is "fixed" it makes further development easier IMO.
 
 @MarkusTeufelberger The pre-commit tool runs the hooks usually (if supported) only on changed files, so it is not that slow after the initial run; the bi advantage being the fixes are suggested before the commit is done and the dev does not need to wait for the CI.
 But I feel adding the "easy" tools would also make a big difference already
 
 @Ajpantuso TBH I would love to have yamllint and editorconfig files from the `ansible/ansible` main project too
 
 @felixfontein the task names are a big part of what made me want to have rules^^ (as in speaking task names that describe more of the test scenario). The second part is the yamllint that's integrated in `ansible-lint` as it stops lines from getting too long to read (forcing the dev to break it down into (hopefully better readable) smaller sections

**Comment by JensHeinrich** [comment]
Regarding security: maybe [bandit](https://github.com/PyCQA/bandit) could also be included?

**Comment by felixfontein** [comment]
yamllint is included in ansible-test and used with (probably) this config: https://github.com/ansible/ansible/blob/devel/test/lib/ansible_test/_util/controller/sanity/yamllint/config/default.yml

> @Ajpantuso TBH I would love to have yamllint and editorconfig files from the `ansible/ansible` main project too

I'm not sure which files you are referring to. I neither see editorconfig files nor a yamllint config in there (except the yamllint configs included in ansible-test).

**Comment by JensHeinrich** [comment]
Yes, that was phrased in a bad way; what I wanted to express is this:

I would like to have an [`.editorconfig](https://editorconfig.org/) file (which defines a few styling settings and is understood by many editors) and a `.yamllint` file (which would be the same as the configuration file you dug up) in the `ansible/ansible` repository so they can be imported in all the community projects.

The yamllint for example would enable everyone to just add the pre-commit hook for yamllint to the repo and that one runs quite fast (so no need to wait for a CI to run it in ansible-test)

**Comment by ssbarnea** [comment]
It should be noted that at some point in the near future, it will no longer be able to upload a new version to galaxy hub if linting is not passing on it. Maybe this would add the extra "motivaction" for addressing this ticket? ;)

PS. I may probably also provide some support with this task, to ease the work.

**Comment by felixfontein** [comment]
@ssbarnea you might want to create an issue on community-topics for that. I think the community and the steering commmittee should have a word on what is allowed to be uploaded to Galaxy and what not.

**Comment by felixfontein** [comment]
Since we now have ansible-lint and code formatting running in CI, I think we can close this.
